File: src/main/java/app/askresume/global/resolver/token/AuthorizationTokenResolver.java
package app.askresume.global.resolver.token;

import app.askresume.global.util.AuthorizationHeaderUtils;
import lombok.RequiredArgsConstructor;
import org.springframework.core.MethodParameter;
import org.springframework.http.HttpHeaders;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.support.WebDataBinderFactory;
import org.springframework.web.context.request.NativeWebRequest;
import org.springframework.web.method.support.HandlerMethodArgumentResolver;
import org.springframework.web.method.support.ModelAndViewContainer;

import javax.servlet.http.HttpServletRequest;

@Component
@RequiredArgsConstructor
public class AuthorizationTokenResolver implements HandlerMethodArgumentResolver {

    @Override
    public boolean supportsParameter(MethodParameter parameter) {
        boolean hasAuthorizationTokenAnnotation = parameter.hasParameterAnnotation(AuthorizationToken.class);
        boolean hasTokenDto = TokenDto.class.isAssignableFrom(parameter.getParameterType());
        return hasAuthorizationTokenAnnotation && hasTokenDto;
    }

    @Override
    public Object resolveArgument(MethodParameter parameter, ModelAndViewContainer mavContainer, NativeWebRequest webRequest, WebDataBinderFactory binderFactory) {
        HttpServletRequest request = (HttpServletRequest) webRequest.getNativeRequest();

        String authorizationHeader = request.getHeader(HttpHeaders.AUTHORIZATION);
        AuthorizationHeaderUtils.validateAuthorization(authorizationHeader);

        String accessToken = authorizationHeader.split(" ")[1];

        return TokenDto.builder()
                .authorizationHeader(authorizationHeader)
                .token(accessToken)
                .build();
    }

}

File: src/main/kotlin/app/askresume/global/resolver/token/AuthorizationTokenResolver.kt
package app.askresume.global.resolver.token

import app.askresume.global.jwt.constant.JwtTokenType
import org.springframework.core.MethodParameter
import org.springframework.stereotype.Component
import org.springframework.web.bind.support.WebDataBinderFactory
import org.springframework.web.context.request.NativeWebRequest
import org.springframework.web.method.support.HandlerMethodArgumentResolver
import org.springframework.web.method.support.ModelAndViewContainer
import javax.servlet.http.HttpServletRequest

@Component
class AuthorizationTokenResolver : HandlerMethodArgumentResolver {
    override fun supportsParameter(parameter: MethodParameter): Boolean {
        val hasAuthorizationTokenAnnotation = parameter.hasParameterAnnotation(AuthorizationToken::class.java)
        val hasTokenDto = TokenDto::class.java.isAssignableFrom(parameter.parameterType)
        return hasAuthorizationTokenAnnotation && hasTokenDto
    }

    override fun resolveArgument(
        parameter: MethodParameter,
        mavContainer: ModelAndViewContainer?,
        webRequest: NativeWebRequest,
        binderFactory: WebDataBinderFactory?,
    ): Any? {
        val request = webRequest.nativeRequest as HttpServletRequest

        val accessTokenCookie = if (request.cookies != null) {
                request.cookies.find { it.name == JwtTokenType.ACCESS.cookieName }
                    ?: throw Exception("미구현 예외") // TODO
            } else {
                throw Exception("미구현 예외") // TODO
            }

        return TokenDto(accessTokenCookie.value)
    }
}
